src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,257 @@
+
+        r = sd_device_get_property_value(device, model_key, model);
+        if (r < 0) {
+                log_error("Model key missing from hwdb for iocost solution '%s'.", name);
+                return -1;
+        }
+
+        r = sd_device_get_property_value(device, qos_key, qos);
+        if (r < 0) {
+                log_error("QoS key missing from hwdb for iocost solution '%s'.", name);
+                return -1;
+        }
+
+        return 0;
+}
+
+static int apply_solution_for_path(const char *path, const char *name) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *model = NULL, *qos = NULL;
+        const char *model_params = NULL, *qos_params = NULL;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+        else if (r == 1) {
+                log_info("Parameters for iocost model missing from hwdb, doing nothing.");
+                return 0;
+        }
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s using solution '%s'\n\tio.cost.model: %s\n\tio.cost.qos: %s\n", path, name ? name : "default", model, qos);
+
+        r = cg_set_attribute("io", NULL, "io.cost.qos", qos);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set qos: %m");
+
+        r = cg_set_attribute("io", NULL, "io.cost.model", model);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set model: %m");
+
+        return 0;
+}
+
+static int query_solutions_for_path(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_strv_free_ char **solutions = NULL;
+        const char *model_name = NULL;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        if ((sd_device_get_property_value(device, "ID_MODEL_FROM_DATABASE", &model_name) < 0 &&
+             sd_device_get_property_value(device, "ID_MODEL", &model_name) < 0)) {
+                log_info("Model name for device %s is unknown", path);
+                model_name = "Unknown";
+        }
+
+        r = get_known_solutions(device, &solutions);
+        if (r < 0)
+                return r;
+        else if (r == 1) {
```

Comment:
drop the `else`, here and everywhere, if the earlier codepaths all exited the function anyway.